File: python/src/themis/policies/quarantine.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Adam Campbell
"""DefaultQuarantinePolicy — the agent-firewall hard stop (RFC v0.2 § 5.4).

A quarantined agent's actions are denied outright until an owner releases it.
The flag may come from ``requestor.metadata["quarantined"]`` (an identified
agent) or ``policy_metadata["quarantine"]["quarantined"]`` (a surface-level
hold). Pure — the store that set the flag is the implementation's business.
"""
from __future__ import annotations

from ..types import Allow, Decision, Deny, PolicyContext


class DefaultQuarantinePolicy:
    name = "quarantine"

    def evaluate(self, ctx: PolicyContext) -> Decision:
        meta = (ctx.policy_metadata or {}).get(self.name) or {}
        rmeta = ctx.requestor.metadata or {}
        if rmeta.get("quarantined") is not True and meta.get("quarantined") is not True:
            return Allow(policy=self.name)
        reason = meta.get("quarantine_reason", rmeta.get("quarantine_reason"))
        return Deny(
            policy=self.name,
            reason="agent_quarantined",
            message="This agent is quarantined. An owner must release it before it can act again.",
            detail={"quarantine_reason": reason},
        )


__all__ = ["DefaultQuarantinePolicy"]

File: packages/core/src/policies/quarantine.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Adam Campbell

/**
 * DefaultQuarantinePolicy — the agent-firewall hard stop (RFC v0.2 § 5.4).
 *
 * A quarantined agent's actions are denied outright until an owner releases
 * it. The flag comes from either place an implementation may carry it:
 *
 *   1. requestor.metadata.quarantined === true          (identified agent)
 *   2. policy_metadata.quarantine.quarantined === true   (surface-level hold)
 *
 * Decision rule: quarantined → deny `agent_quarantined`; otherwise allow.
 * Pure — the store that set the flag is the implementation's business.
 */

import type { IPolicy, IPolicyContext, IPolicyDecision, IQuarantineMetadata } from '../types.js';

export class DefaultQuarantinePolicy implements IPolicy {
  readonly name = 'quarantine' as const;

  async evaluate(ctx: IPolicyContext): Promise<IPolicyDecision> {
    const meta = ctx.policy_metadata?.[this.name] as IQuarantineMetadata | undefined;
    const fromRequestor = ctx.requestor.metadata?.quarantined === true;
    const fromMeta = meta?.quarantined === true;
    if (!fromRequestor && !fromMeta) {
      return { kind: 'allow', policy: this.name };
    }
    const reason = (meta?.quarantine_reason ?? (ctx.requestor.metadata?.quarantine_reason as string | undefined)) ?? null;
    return {
      kind: 'deny',
      policy: this.name,
      reason: 'agent_quarantined',
      message: 'This agent is quarantined. An owner must release it before it can act again.',
      detail: { quarantine_reason: reason },
    };
  }
}
